cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -443,8 +443,11 @@ struct ScalarToProtoImpl {
     return Status::OK();
   }
 
+  // Note: while std::string&& would be better for the argument
+  // type, older versions of protobuf don't have this overload
+  // yet.
   template <typename ScalarWithBufferValue>
-  Status FromBuffer(void (substrait::Expression::Literal::*set)(std::string&&),
+  Status FromBuffer(void (substrait::Expression::Literal::*set)(const std::string&),
```

Comment:
You could also take a lambda and let C++ resolve the best overload there:
```c++
  template <typename LitteralSetter>
  Status FromBuffer(LitteralSetter&& set_lit,
                    const ScalarWithBufferValue& scalar_with_buffer) {
    set_lit(lit_, scalar_with_buffer.value->ToString());
    return Status::OK();
  }

  Status Visit(const StringScalar& s) {
    return FromBuffer([](Lit* lit, std::string s) { lit->set_string(std::move(s)); });
  }
```

Reply:
That yields some positively ugly code IMO, but here it is: 7ffd90f

Reply:
If the lambda is too ugly, we can also remove the `FromBuffer` helpers, which will add at most one line of code to each caller. In both cases the bottom line is that we don't hardcode the actual protobuf method signature, so we're more robust against signature changes.